File: 02-observer/weather-node/displays/stats.ts
import Subject from "../weatherdata/subject";
import DisplayElement from "./display";
import Observer from "./observer";

export default class StatisticsDisplay implements Observer, DisplayElement {
  private maxTemp: number;
  private minTemp: number;
  private tempSum: number;
  private numReadings: number;
  private weatherData: Subject;

  constructor(weatherData: Subject) {
    this.weatherData = weatherData;
    this.weatherData.registerObservers(this);
    this.maxTemp = 0;
    this.minTemp = 200;
    this.tempSum = 0;
    this.numReadings = 0;
  }

  public update(temp: number, humidity: number, pressure: number): void {
    this.tempSum += temp;
    this.numReadings++;

    if (temp > this.maxTemp) {
      this.maxTemp = temp;
    }

    if (temp < this.minTemp) {
      this.minTemp = temp;
    }

    this.display();
  }

  public display(): void {
    console.log(
      `Avg/Max/Min temperatures = ${this.tempSum / this.numReadings}/${
        this.maxTemp
      }/${this.minTemp}`
    );
  }
}

File: 02-observer/weather-go/display/stats.go
package display

import (
	"fmt"

	"github.com/k-nox/design-patterns/02-observer/weather-go/observer"
)

type Statistics struct {
	maxTemp float32
	minTemp float32
	tempSum float32
	numReadings int
	weatherData observer.Subject
}

func (s *Statistics) Update(temp float32, humidity float32, pressure float32) {
	s.tempSum += temp;
	s.numReadings++;
	if temp > s.maxTemp {
		s.maxTemp = temp;
	}

	if temp < s.minTemp {
		s.minTemp = temp;
	}

	s.Display()
}

func (s *Statistics) Display() {
	fmt.Printf("Avg/Max/Min temperatures: %.2f/%.2f/%.2f\n", s.tempSum / float32(s.numReadings), s.maxTemp, s.minTemp)
}

func NewStats(wd observer.Subject) *Statistics {
	s := &Statistics{
		minTemp: 200,
		weatherData: wd,
	}

	wd.RegisterObserver(s)
	return s
}